firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/RingBuffer.txt

```diff
@@ -0,0 +1,592 @@
+    // internal state should look like [][*^][][][]
+    WHEN('2 bytes of data are pushed ') 
+      THEN('The push method returns ok status') 
+      THEN('After the push method, peek method returns expected data') 
+        // internal state should look like [][0*][1][][]
+
+    WHEN('1 byte of data is poped ') 
+      THEN('The pop method returns empty status') 
+
+  GIVEN('A RingBuffer with capacity 4 to which 2 bytes of data were pushed and 1 byte was poped')   // [][1*][2][3^][]
+    // internal state should look like [][1*][^][][]
+    WHEN('1 byte of data is pushed ') 
+      THEN('The push method returns ok status') 
+      THEN('After the push method, peek method returns expected data') 
+        // internal state should look like [][1*][2][^]
+
+    WHEN('1 bytes of data are poped ') 
+      THEN('The pop method returns ok status') 
+      THEN('After the pop method,the peek method returns expected data in RingBuffer') 
+        // internal state should look like [][*^][][][]
+
+  GIVEN('A RingBuffer with capacity 4 to which 3 bytes of data were pushed and 1 byte was poped')   // [][][2*][3][4^]
+    // internal state should look like [][1*][2][^][]
+    WHEN('2 bytes of data are pushed ') 
+      THEN('The push method returns ok status') 
+      THEN('After the push method, peek method returns expected data') 
+        // internal state should look like [^][1*][2][3][4]
+
+    WHEN('1 byte of data is poped ') 
+      THEN('The pop method returns ok status') 
+      THEN('After the pop method,the peek method returns expected data in RingBuffer') 
+        // internal state should look like [][][2*][^][]
+
+  GIVEN('A RingBuffer with capacity 4 to which 4 bytes of data were pushed and 1 byte was poped')   // [0][1^][][][4*]
+    // internal state should look like [][1*][2][3][^]
+    WHEN('1 bytes of data are pushed ') 
+      THEN('The push method returns ok status') 
+      THEN('After the push method, peek method returns expected data') 
+        // internal state should look like [^][1*][2][3][4]
+
+    WHEN('1 byte of data is poped ') 
+      THEN('The pop method returns ok status') 
+      THEN('After the pop method,the peek method returns expected data in RingBuffer') 
+        // internal state should look like [][][2*][3][^]
+
+  GIVEN('A RingBuffer with capacity 4 to which 2 bytes of data were pushed and 1 byte was poped and 3 bytes were pushed')   // [0^][1*][][][]
+    // internal state should look like [^][1*][2][3][4]
+    WHEN('2 bytes of data are pushed ') 
+      THEN('The push method returns full status') 
+      THEN('After push method, peek method returns expected data') 
+        // internal state should look like [^][1*][2][3][4]
+
+    WHEN('1 byte of data is poped ') 
+      THEN('After the pop method,the peek method returns the data in RingBuffer') 
+        // internal state should look like [^][][2*][3][4]
```

Comment:
As another example of THENs we need in order to break down "internal state should look like [^][][2*][3][4]" into concrete THENs: after the pop method call, it should be possible to push one more byte onto the ring buffer, but it should not be possible to push more than one byte onto the ring buffer